### PR TITLE
Rework RosettaStone structure

### PR DESCRIPTION
**Pull request by utilForever**
This revision includes:
- Rework entity system (#345)
    - Create class `Playable`: This class is an entity which can be played from hand zone into play zone. In other words, the entity can be played.
    - Class `Player` inherits from class `Entity`.
    - Class `Entity` has variable `game`, `player` to access directly.
- Rework zone system (#346)
    - Classes related class `IZone` stores object `Playable` instead of `Entity`.
- Rework targeting system (#347)
    - Create and move methods in class `Entity` to `Playable`: `IsPlayable()`, `TargetingRequirements()`, `IsPlayableByPlayer()`, `IsPlayableByCardReq()`, `IsValidPlayTarget()`, `HasAnyValidPlayTargets()`.
    - Remove all functions in `Targeting.hpp` and its file.
    - Create methods in class `Card` to check a target is valid.
        - `TargetingRequirements()`: Calculates if a target is valid by testing the game state for each hardcoded requirement.
        - `GetValidPlayTargets()`: Gets the valid play targets. This method defaults to targeting in the context of spells/hero powers.
    - Create class `TargetingPredicates`: This class includes utility methods for targeting predicate.
    - Now, perform the processing related to targeting in `Card::Initialize()` such as targeting predicate and targeting type.
- Rework aura system (#342) and `AdaptiveEffect` (#344)
    - Create aura related classes and structs
        - Class `IAura`: This class is interface for effects of enchantments that should be updated during Game::UpdateAura().
        - Struct `AuraUpdateInstruction`: This struct stores source and instruction data for updating aura.
        - Class `Aura`: This class represent aura. Auras can affect entities and change the applied entities attack, cost and so on. Aura must be activated first to affect entities.
        - Class `AdaptiveCostEffect`: Implementation of the specific effects of varying cost. e.g. Giants.
        - Class `AdaptiveEffect`: Effects of this kind of Auras are influenced by other factors in game, in real time. e.g. Lightspawn, Southsea Deckhand.
        - Class `AdjacentAura`: This class is an aura for adjacent type. Note that some auras only affect djacent minions, e.g. Dire Wolf Alpha and Flametongue Totem.
        - Class `EnrageEffect`: Enrage is the informal name for a minion ability where the stated effect becomes active only when the minion is damaged. If the minion regains undamaged status, either through healing or a forced resetting of its maximum Health, the Enrage effect will be canceled.
        - Class `SwitchingAura`: This class is an aura that can switch between on and off. e.g. Pint-Sized Summoner and Kalecgos.
- Rework effect system (#343) and `AuraEffects` (#344)
    - Create attribute related classes
        - Class `Attr`: This class is an interface for attribute.
        - Class `IntAttr`: This class inherits from Attr class and is template specialization for the attribute that is integer type.
        - Class `BoolAttr`: This class inherits from Attr class and is template specialization for the attribute that is boolean type.
        - Class `SelfContainedIntAttr`: This class inherits from IntAttr class and contains Effect() method for generating generic effect using CRTP(Curiously Recurring Template Pattern) technique.
        - Class `SelfContainedBoolAttr`: This class inherits from BoolAttr class and contains Effect() method for generating generic effect using CRTP(Curiously Recurring Template Pattern) technique.
        - Class `Atk`: This class is an attribute for attack and inherits from SelfContainedIntAttr class. It uses CRTP(Curiously Recurring Template Pattern) technique.
        - Class `Cost`: This class is an attribute for cost and inherits from SelfContainedIntAttr class. It uses CRTP(Curiously Recurring Template Pattern) technique.
        - Class `Health`: This class is an attribute for health and inherits from SelfContainedIntAttr class. It uses CRTP(Curiously Recurring Template Pattern) technique.
    - Create effect related classes
        - Class `IEffect`: This class is an interface for effect.
        - Class `GenericEffect`: This class is a simple container for saving game tag such as GameTag::ATK, GameTag::COST and GameTag::HEALTH.
    - Remove class `Aura`
- Rework task system
    - Create class `PriorityQueue`: This class is a custom priority queue implementation for convenience.
    - Create struct `EventMetaData`: This struct is temporary meta data for event.
    - Create class 'CostManager`: This class manages the cost of the card. It is affected by cost aura, adaptive effect and enchantment.
- Implement 6 cards in EXPERT1 cardset
    - Mountain Giant (EX1_105)
    - Betrayal (EX1_126)
    - Nobel Sacrifice (EX1_130)
    - Eye for an Eye (EX1_132)
    - Redemption (EX1_136)
    - Headcrack (EX1_137)
- Change the logic of some cards
    - Pint-Sized Summoner (EX1_076)
    - Commanding Shout (NEW1_036e2)
    - Alarm-o-Bot (EX1_006)
    - Arcane Missiles (EX1_277)
- Move 2 'CORE' cards to 'HoF` due to Hearthstone update
    - Mind Blast (DS1_233)
    - Vanish (NEW1_004)
- Add/Change/Remove some tasks
    - Add: `ChangeAttackingTargetTask`, `GetEventNumberTask`, `IncludeAdjacentTask`, `MathNumberIndexTask`, `MoveToSetasideTask`, `SummonOpTask`
    - Change: `FuncEntityTask` → `FuncPlayableTask`
    - Remove: `MathSubTask`
- Create trigger function to process `TriggerType::DEATH`
- Update `pybind11` library version to v2.4.3

### CODE REVIEW

File: Includes/Rosetta/Commons/Utils.hpp [hunk 1]
```diff
@@ -55,6 +66,53 @@ void EraseIf(ContainerT& items, const PredicateT& predicate)
     }
 }
 
+//! Gets N elements from a list of distinct elements by using the default
+//! equality comparer. The source list must not have any repeated elements.
+//! \param list A list of distinct elements to choose.
+//! \param amount The number of elements to choose.
+//! \return A list of N distinct elements.
+template <typename T>
+std::vector<T*> ChooseNElements(const std::vector<T*>& list, int amount)
```
Comment (CreatorSeraph) [inline Includes/Rosetta/Commons/Utils.hpp]: amount의 자료형을 size_t나 unsigned int로 하는건 어떤가요?

File: Includes/Rosetta/Enchants/Attrs/Cost.hpp [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+
+    //! Applies the effect that affects the attribute.
+    //! \param entity The entity to apply the effect.
+    //! \param effectOp The effect operator to change the attribute.
+    //! \param value The value to change the attribute.
+    void Apply(Entity* entity, EffectOperator effectOp, int value) override
+    {
+        SelfContainedIntAttr::Apply(entity, effectOp, value);
+
+        const auto playable = dynamic_cast<Playable*>(entity);
```
Comment (CreatorSeraph) [inline Includes/Rosetta/Enchants/Attrs/Cost.hpp]: 이런경우 static_cast를 사용해도 괜찮지 않을까요?

File: Includes/Rosetta/Commons/Utils.hpp (hunk 1 above)
Comment (utilForever) [reply Includes/Rosetta/Commons/Utils.hpp]: 수정하겠습니다.

File: Includes/Rosetta/Enchants/Attrs/Cost.hpp (hunk 2 above)
Comment (utilForever) [reply Includes/Rosetta/Enchants/Attrs/Cost.hpp]: C++ Core Guidelines에 보면 Base 클래스에서 Derived 클래스로 다운 캐스트 할 때는 `static_cast`를 쓰지 마라는 지침이 있어서 이를 참고했습니다.

**Comment by utilForever** [comment]
I'll ignore test coverage in this PR because it's a task that changes the overall structure.

File: Includes/Rosetta/Enchants/Attrs/IntAttr.hpp [hunk 3]
```diff
@@ -0,0 +1,169 @@
...
+
+    //! Applies the effect that affects the attribute.
+    //! \param entity The entity to apply the effect.
+    //! \param effectOp The effect operator to change the attribute.
+    //! \param value The value to change the attribute.
+    void Apply(T* entity, EffectOperator effectOp, int value) override
+    {
+        const int target = GetValue(entity);
+
+        if (effectOp == EffectOperator::SET)
```
Comment (1cekrim) [inline Includes/Rosetta/Enchants/Attrs/IntAttr.hpp]: 아래 switch 문에 집어넣어도 괜찮지 않을까요?

File: Sources/Rosetta/Actions/Generic.cpp [hunk 4]
```diff
@@ -5,82 +5,119 @@
...
+        if (auto aura = power.GetAura(); aura)
+        {
+            aura->Activate(enchantment);
+        }
+
+        if (auto trigger = power.GetTrigger(); trigger)
+        {
+            trigger->Activate(enchantment);
+        }
+    }
```
Comment (1cekrim) [inline Sources/Rosetta/Actions/Generic.cpp]: 47-48에서 GetAura()와 GetRigger()가 nullptr인지 검사했는데, 55와 60에서도 이 둘이 nullptr인지 검사하고 있습니다.
```
if (!power.GetDeathrattleTask().empty())
{
    // Create Enchantment instance Only when it is needed.
    // As an owner entity for auras, triggers or deathrattle tasks.
    Enchantment* enchantment =
        Enchantment::GetInstance(creator->player, enchantmentCard, target);
}
if (auto aura = power.GetAura(); aura)
{
    aura->Activate(enchantment);
}

if (auto trigger = power.GetTrigger(); trigger)
{
    trigger->Activate(enchantment);
}
```
이렇게 바꿔도 될 것 같습니다.

File: Includes/Rosetta/Enchants/Attrs/IntAttr.hpp (hunk 3 above)
Comment (utilForever) [reply Includes/Rosetta/Enchants/Attrs/IntAttr.hpp]: switch문 안에 넣는 건 스코프에 맞지 않고 바로 위로 옮기는 건 가능합니다.

Comment (utilForever) [reply Includes/Rosetta/Enchants/Attrs/IntAttr.hpp]: 제가 잘못 봤네요. 무슨 말인지 이해했고 반영해서 수정하겠습니다.

**Comment by utilForever** [comment]
I'll complete this PR and merge it now. Subsequent code reviews will be reflected as the trigger system is reworked. Thank you all for your trouble! :tada:

**Pull request merged**